### PR TITLE
Add `cookieJar` option

### LINKED ISSUE #566: Add native support for cookie persistence handling using an API compatible with tough-cookie
Adding native cookie support has been discussed in multiple issues, e.g.

* https://github.com/sindresorhus/got/issues/107

However, none of the issues cover how one is supposed to implement cookie persistence between redirects.

It is simple to implement something like [tough-cookie](https://www.npmjs.com/package/tough-cookie) for a single request with no redirects. However, I haven't figured out how to keep track of cookies when a cookie is set on a redirect and the final destination depends on the cookie being present. It probably requires handling "response" and "redirect" events and modifying the request. Point is, as a consumer of an HTTP client, I want an in-built cookie support or at least an API and a recipe that allows to quickly add cookie support. Otherwise, the learning curve for starting to use Got becomes unreasonably high compared to the available alternatives, all of which (request, axios, superagent) support cookie persistence out of the box.

### PR DESCRIPTION
**Pull request by szmarczak**
Fixes #566

 - [x] Works
 - [x] Documentation
 - [x] Tests for invalid cookies

### CODE REVIEW

File: source/normalize-arguments.js [hunk 1]
```diff
@@ -1,8 +1,5 @@
 'use strict';
-/* istanbul ignore next: compatibility reason */
-const URLGlobal = typeof URL === 'undefined' ? require('url').URL : URL; // TODO: Use the `URL` global when targeting Node.js 10
-/* istanbul ignore next: compatibility reason */
-const URLSearchParamsGlobal = typeof URLSearchParams === 'undefined' ? require('url').URLSearchParams : URLSearchParams;
+const {URL, URLSearchParams} = require('url'); // TODO: Use the `URL` global when targeting Node.js 10
```
[the remark this answers is not in the document]
Comment (szmarczak) [reply source/normalize-arguments.js]: It's more readable.

[the remark this answers is not in the document]
Comment (sindresorhus) [reply source/normalize-arguments.js]: Yes, but it's not relevant to this PR. Just do a commit on master instead.

[the remark this answers is not in the document]
Comment (szmarczak) [reply source/normalize-arguments.js]: K

File: test/cookies.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+	await s.listen(s.port);
+});
+
+test.after('cleanup', async () => {
+	await s.close();
+});
+
+test('reads a cookie', async t => {
+	const jar = new tough.CookieJar();
```
Comment (sindresorhus) [inline test/cookies.js]: `jar` => `cookieJar`

**Comment by sindresorhus** [comment]
@gajus Are you happy with this?

File: readme.md [hunk 3]
```diff
@@ -161,9 +161,11 @@ The `content-length` header will be automatically set if `body` is a `string` /
 
 ###### cookieJar
 
-Type: `object`
+Type: [`tough.CookieJar` instance](https://github.com/salesforce/tough-cookie#cookiejar)
 
-An instance of [`tough.CookieJar`](https://github.com/salesforce/tough-cookie#cookiejar). If provided, `options.headers.cookie` will be overriden.
+Supports cookies globally. You don't need to care of parsing nor storing them. [Click here](#cookies) for an example.
+
+**Note**: if provided, `options.headers.cookie` will be overriden.
```
Comment (gajus) [inline readme.md]: I would throw an error if cookie header is provided and `cookieJar` is configured.

Comment (szmarczak) [reply readme.md]: Good point!

Comment (szmarczak) [reply readme.md]: Hmmm. The `cookie` header should be taken if it exists.

File: readme.md [hunk 4]
```diff
@@ -159,6 +159,14 @@ If present in `options` and `options.method` is not set, `options.method` will b
 
 The `content-length` header will be automatically set if `body` is a `string` / `Buffer` / `fs.createReadStream` instance / [`form-data` instance](https://github.com/form-data/form-data), and `content-length` and `transfer-encoding` are not manually set in `options.headers`.
 
+###### cookieJar
+
+Type: [`tough.CookieJar` instance](https://github.com/salesforce/tough-cookie#cookiejar)
+
+Supports cookies globally. You don't need to care of parsing nor storing them. [Click here](#cookies) for an example.
```
Comment (sindresorhus) [inline readme.md]: `Click here` kinda links is a bad pattern. Let's make it descriptive instead. Here, it can just be `[Example.]()`.

Comment (szmarczak) [reply readme.md]: Yeah, true.

File: readme.md (hunk 3 above)
Comment (szmarczak) [reply readme.md]: Ok then

Comment (szmarczak) [reply readme.md]: done

**Comment by sindresorhus** [comment]
💯 👍

**Pull request merged**